File: stack.js
// ES2015

class Stack {

    constructor(length) {
        this.items = new Array(length);
        this.index = 0;
    }

    isEmpty() {
        return this.index === 0;
    }

    isFull() {
        return this.index === this.items.length;
    }

    peek() {

        if (this.isEmpty()) {
            throw new RangeError("Cannot peek at an empty stack!");
        }

        return this.items[this.index - 1];
    }

    pop() {

        if (this.isEmpty()) {
            throw new RangeError("Cannot pop from an empty stack!");
        }

        this.index -= 1;
        let item = this.items[this.index];
        this.items[this.index] = undefined;
        return item;
    }

    push(item) {

        if (this.isFull()) {
            throw new RangeError("Cannot push into a full stack!");
        }

        this.items[this.index] = item;
        this.index += 1;
    }

    size() {
        return this.index;
    }

}

File: Stack/stack.go

package stack

import "errors"

type Stack struct {
	Items []int
	Index int
}

func NewStack(l int) *Stack {
	return &Stack{Items: make([]int, l, l*2), Index: 0}
}

func (s *Stack) Empty() bool {
	return s.Index == 0
}

func (s *Stack) Full() bool {
	return s.Index == len(s.Items)
}

func (s *Stack) Peek() (int, error) {

	if s.Empty() {
		return 0, errors.New("Cannot peek at an empty stack!")
	}

	return s.Items[s.Index], nil
}

func (s *Stack) Pop() (int, error) {

	if s.Empty() {
		return 0, errors.New("Cannot pop from an empty stack!")
	}

	s.Index -= 1
	i := s.Items[s.Index]
	s.Items = s.Items[:s.Index]
	return i, nil
}

func (s *Stack) Push(i int) error {

	if s.Full() {
		errors.New("Cannot push into a full stack!")
	}

	s.Items = append(s.Items, i)
	s.Index += 1
	return nil
}

func (s *Stack) Len() int {
	return s.Index
}
